TextEnvironments

lvwerra opened the pull request:
This PR adds multi-turn text environment to TRL. 

Target API

```python
env = Environment(model, tokenizer, prompt, tools)

for tasks in ppo_trainer.dataloader:
    histories = env.run(tasks)
    tokens, mask = histories.get_tokens()
    ppo_trainer.step(tokens, mask=mask, histories.rewards)
    
# alternatively, this would probably be much nicer
for tasks in ppo_trainer.dataloader:
    tokens, masks, rewards, history = env.run(tasks)
    ppo_trainer.step(tokens, masks, rewards)
```

Todos
- [x] text to token+mask conversion | #429 
- [ ] penalty for tool calls
- [x] masking in `PPOTrainer` after fwd pass | #429 
- [x] end-to-end example calculater

Current working example

```python
from trl import TextEnvironment, TextHistory, AutoModelForCausalLMWithValueHead
from transformers import AutoModelForCausalLM, AutoTokenizer, load_tool

tool = load_tool("ybelkada/simple-calculator")

model_id = "gpt2-xl"

model = AutoModelForCausalLMWithValueHead.from_pretrained(model_id)
tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token

prompt = """\
What is 12.1 + 1 - 3?
<request><SimpleCalculatorTool>12.1 + 1<call>13.1<response>
<request><SimpleCalculatorTool>13.1 - 3<call>10.1<response>
Result = 10.1 <submit>

"""

reward_fn = lambda x: 1

env = TextEnvironment(model, tokenizer,[tool], reward_fn, prompt, generation_kwargs={"max_new_tokens": 32})
h = env.run(["What is 387 * 228?"])

h[0].show()
```

Result:
```text
What is 12.1 + 1 - 3?
<request><SimpleCalculatorTool>12.1 + 1<call>13.1<response>
<request><SimpleCalculatorTool>13.1 - 3<call>10.1<response>
Result = 10.1 <submit>

What is 387 * 228?

<request><SimpleCalculatorTool>387 * 228<call>88236.0<response>

Result = 88236.0 <submit>
Reward: 1
```

Where the `88236.0<response>` segement was generated by the tool call.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/trl/pr_424). All of your documentation changes will be reflected on that endpoint.

lvwerra commented:
Did a small refactor and added a bit of docs. A few things that are left to do:
- sort out the examples. we have three: calculator/search/python. should we move them to research examples?
- remove gradio demo again - i think it should only live on the hub where it is also version controlled
- @vwxyzjn can you move the images to https://huggingface.co/trl-internal-testing (see e.g. the text env docs i added). this will make the repo less bloated with images

In any case I think it is ready for a first review modulo the points above @younesbelkada & @vwxyzjn

younesbelkada commented on examples/triviaqa.py:
```diff
@@ -0,0 +1,159 @@
+        yield dataset[i]["question"], [item for item in dataset[i]["answer"]["normalized_aliases"]]
+gen = data_generator()
+gen = iter(gen)
+# dataset[0]["question"]
+# dataset[0]["answer"]['normalized_aliases']
```
```suggestion
```

younesbelkada commented on examples/tool_demo.py:
```diff
@@ -0,0 +1,350 @@
+
+
+HF_TOKEN = os.environ.get("HF_TOKEN", None)
+print(HF_TOKEN)
```
```suggestion
```

younesbelkada commented on docs/source/static/learning_tools.png:
Images to move in `trl-internal-testing`, here: https://huggingface.co/datasets/trl-internal-testing/example-images

vwxyzjn commented:
Thanks @younesbelkada for the comments, I have addressed most of them. @lvwerra the PR looks good. Quick question: do we want to merge `docs/source/learning_tools.mdx` and `docs/source/text_environments.md`?

lvwerra commented:
> Quick question: do we want to merge docs/source/learning_tools.mdx and docs/source/text_environments.md?

I thought we could have `text_environments.md` as the basic doc for how TextEnvs work and the `learning_tools.mdx` as the more hands-on guide also linking the experiment scripts. Wdyt?

Also @vwxyzjn would you mind fixing the scripts so they pass the quality checks. In my opinion we can also exclude them from the quality tests, no strong opinion.

lvwerra commented on trl/environment/base_environment.py:
```diff
@@ -0,0 +1,460 @@
+    from rich import print
+    from rich.text import Text
+except ImportError:
+    is_rich_available = False
```
done!

lvwerra commented on trl/environment/base_environment.py:
```diff
@@ -0,0 +1,460 @@
+import re
```
done!

lvwerra commented on trl/trainer/ppo_trainer.py:
```diff
@@ -539,6 +541,8 @@ def _step_safety_checker(
             scores (List[`torch.FloatTensor`]):
                 List of tensors containing the scores.
+            masks (List[`torch.LongTensor`], *optional*):
+                list of optional tensors containing the masks of shape (`query_length` + `response_length`)
```
added the same tests as for the others

lvwerra commented on docs/source/learning_tools.mdx:
```diff
@@ -0,0 +1,229 @@
+Using Large Language Models (LLMs) with tools has been a popular topic recently with awesome works such as [ToolFormer](https://arxiv.org/abs/2302.04761) and [ToolBench](https://arxiv.org/pdf/2305.16504.pdf). In TRL, we provide a simple example of how to teach LLM to use tools with reinforcement learning. 
+
+
+Here's an overview of the scripts in the [trl repository](https://github.com/lvwerra/trl/tree/main/examples/research_projects):
+
+| File | Description | 
+|---|---| 
+| [`calculator.py`](https://github.com/lvwerra/trl/blob/main/examples/research_projects/calculator.py) | Script to train LLM to use a calculator with reinforcement learning. |
+| [`triviaqa.py`](https://github.com/lvwerra/trl/blob/main/examples/research_projects/triviaqa.py) | Script to train LLM to use a wiki tool to answer questions. |
+| [`python_interpreter.py`](https://github.com/lvwerra/trl/blob/main/examples/research_projects/python_interpreter.py) | Script to train LLM to use python interpreter to solve math puzzles. |
```
```suggestion
Here's an overview of the scripts in the [trl repository](https://github.com/lvwerra/trl/tree/main/examples/research_projects/tools):

| File | Description | 
|---|---| 
| [`calculator.py`](https://github.com/lvwerra/trl/blob/main/examples/research_projects/tools/calculator.py) | Script to train LLM to use a calculator with reinforcement learning. |
| [`triviaqa.py`](https://github.com/lvwerra/trl/blob/main/examples/research_projects/tools/triviaqa.py) | Script to train LLM to use a wiki tool to answer questions. |
| [`python_interpreter.py`](https://github.com/lvwerra/trl/blob/main/examples/research_projects/tools/python_interpreter.py) | Script to train LLM to use python interpreter to solve math puzzles. |
```

younesbelkada commented on examples/research_projects/tools/calculator.py:
```diff
@@ -0,0 +1,104 @@
+import re
```
```suggestion
# coding=utf-8
# Copyright 2023 The HuggingFace Inc. team. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import re
```

younesbelkada commented on examples/research_projects/tools/python_interpreter.py:
```diff
@@ -0,0 +1,179 @@
+import os
```
```suggestion
# coding=utf-8
# Copyright 2023 The HuggingFace Inc. team. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
```

younesbelkada commented on examples/research_projects/tools/triviaqa.py:
```diff
@@ -0,0 +1,174 @@
+import os
```
```suggestion
# coding=utf-8
# Copyright 2023 The HuggingFace Inc. team. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
```

younesbelkada commented on trl/trainer/ppo_trainer.py:
```diff
@@ -583,13 +589,20 @@ def step(
         bs = self.config.batch_size
 
-        queries, responses, scores = self._step_safety_checker(bs, queries, responses, scores)
+        # if response_masks is None:
+        #    response_masks = [torch.ones_like(response) for response in responses]
```
```suggestion
```

younesbelkada reviewed (approved):
Looking great ! Thanks for all your efforts, left 4 nits, otherwise LGTM !
